File: src/01_basics/basics_py/basics_py/action_client.py
"""
TUTORIAL 3b — Actions: Client

Sends a goal to /count_up and prints feedback + result.

    ros2 run basics action_server    (separate terminal first)
    ros2 run basics action_client
"""

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from example_interfaces.action import Fibonacci


class CountClient(Node):
    def __init__(self):
        super().__init__('count_client')

        # === CHECKPOINT: actions ===
        self._client = ActionClient(self, Fibonacci, 'count_up')
        # === END CHECKPOINT: actions ===

    # === CHECKPOINT: actions ===
    def send_goal(self, target: int):
        self._client.wait_for_server()
        self.get_logger().info(f'Sending goal: count to {target}')

        goal = Fibonacci.Goal()
        goal.order = target

        future = self._client.send_goal_async(goal, feedback_callback=self.on_feedback)
        future.add_done_callback(self.on_goal_accepted)

    def on_feedback(self, feedback_msg):
        count = feedback_msg.feedback.sequence[0]
        self.get_logger().info(f'Feedback: {count}')

    def on_goal_accepted(self, future):
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().warn('Goal rejected')
            return
        self.get_logger().info('Goal accepted — waiting for result...')
        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(self.on_result)

    def on_result(self, future):
        result = future.result().result
        self.get_logger().info(f'Result: reached {result.sequence[0]}')
        rclpy.shutdown()
    # === END CHECKPOINT: actions ===


def main(args=None):
    rclpy.init(args=args)
    node = CountClient()

    # === CHECKPOINT: actions ===
    node.send_goal(5)
    # === END CHECKPOINT: actions ===

    rclpy.spin(node)
    node.destroy_node()

File: src/01_basics/basics_cpp/src/action_client.cpp
// TUTORIAL 3b — Actions: Client
//
// ros2 run basics_cpp action_server    (separate terminal first)
// ros2 run basics_cpp action_client

#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "example_interfaces/action/fibonacci.hpp"

using Fibonacci = example_interfaces::action::Fibonacci;
using GoalHandle = rclcpp_action::ClientGoalHandle<Fibonacci>;

class CountClient : public rclcpp::Node
{
public:
  CountClient() : Node("count_client")
  {
    // === CHECKPOINT: actions ===
    client_ = rclcpp_action::create_client<Fibonacci>(this, "count_up");
    // === END CHECKPOINT: actions ===
  }

  // === CHECKPOINT: actions ===
  void send_goal(int target)
  {
    client_->wait_for_action_server();
    RCLCPP_INFO(get_logger(), "Sending goal: count to %d", target);

    auto goal = Fibonacci::Goal();
    goal.order = target;

    auto opts = rclcpp_action::Client<Fibonacci>::SendGoalOptions();

    opts.feedback_callback = [this](GoalHandle::SharedPtr,
      const std::shared_ptr<const Fibonacci::Feedback> fb)
    {
      RCLCPP_INFO(get_logger(), "Feedback: %d", fb->sequence[0]);
    };

    opts.result_callback = [this](const GoalHandle::WrappedResult & res)
    {
      RCLCPP_INFO(get_logger(), "Result: reached %d", res.result->sequence[0]);
      rclcpp::shutdown();
    };

    client_->async_send_goal(goal, opts);
  }

private:
  rclcpp_action::Client<Fibonacci>::SharedPtr client_;
  // === END CHECKPOINT: actions ===
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CountClient>();

  // === CHECKPOINT: actions ===
  node->send_goal(5);
  // === END CHECKPOINT: actions ===

  rclcpp::spin(node);
}
